### PR TITLE
#77 - no need to have the groups ordered

### PR DESCRIPTION
**Pull request by arjantijms**
The constraint to iterate the stores to obtain groups in order is not necessary. They simply need to be collected all, and the final result is a set without any order.

### CODE REVIEW

**Comment by arjantijms** [comment]
See issue: https://github.com/javaee-security-spec/soteria/issues/77

**Comment by wmhopkins** [comment]
Are we sure we want to remove this requirement? It's true that the resulting set of groups is not ordered, and that therefore order of calling isn't important for ordering the returned groups, but we aren't talking about what all IdentityStoreHandlers must do, we're only talking about the required behavior of the default handler, and there might be reasons why ordering could matter. For one thing, it simplifies the logic, since the idstores were already ordered when looking for authentication stores; for another, order might be important for other reasons -- perhaps one idstore is coded to only fetch groups if no groups were found by a previous idstore, and is configured with a high number, to run last? At minimum, specifying the order provides predictability and prevents implementations from diverging. Ordering doesn't seem necessary, but it seems harmless, and contributes to well-defined behavior.

Just thinking out loud -- I'm OK to remove if there is consensus in the EG.

**Comment by arjantijms** [comment]
It would not be bad if they (identity stores) were consulted in an ordered way, but it doesn't matter that much as eventually all stores needs to be called anyway.

It would make a little bit of difference if for a given set of stores one fails, and several slow ones don't. If the slow ones happen to have a higher order it would fail faster in the ordered way, and more randomly otherwise.

Fetching groups is not in the path where authentication itself can still fail, so I don't think this matters for timing attacks. (if the group fetching fails it's almost certainly a system level error).

tldr; I too am okay with either leaving it in or removing it.

**Comment by rdebusscher** [comment]
Since always all Identity stores (related to retrieving the groups) are
consulted and results are aggregated in a Set, I don't see the advantages
of having a specific order.

So I don't see any reason why we need to put that sentence in the spec.

But since the Identity stores are already ordered, it feels natural to use
that order.

Rudy

**Comment by wmhopkins** [comment]
I have a slight preference for leaving the ordering in place, simply for predictability of behavior, but I'm also OK either way.

@arjantijms, now that you've heard the counter-arguments, let me know what you want to do. I can merge the PR or close it.

**Comment by arjantijms** [comment]
Okay, well, then close the PR. It's not such a big deal. thx!
